components/eamxx/src/diagnostics/vertical_layer.cpp

```diff
@@ -0,0 +1,124 @@
+  const auto npacks_p1 = ekat::npack<Pack>(m_num_levs+1);
+  const auto npacks    = ekat::npack<Pack>(m_num_levs);
+  m_z_int = view_2d("z_int",m_num_cols,npacks_p1);
+  m_z_mid = view_2d("z_mid",m_num_cols,npacks);
+}
+// =========================================================================================
+void VerticalLayerDiagnostic::compute_diagnostic_impl()
+{
+  const auto npacks         = ekat::npack<Pack>(m_num_levs);
+  const auto default_policy = ekat::ExeSpaceUtils<KT::ExeSpace>::get_thread_range_parallel_scan_team_policy(m_num_cols, npacks);
+
+  const auto& phis               = get_field_in("phis").get_view<const Real*>();
+  const auto& T_mid              = get_field_in("T_mid").get_view<const Pack**>();
+  const auto& p_mid              = get_field_in("p_mid").get_view<const Pack**>();
+  const auto& qv_mid             = get_field_in("qv").get_view<const Pack**>();
+  const auto& pseudo_density_mid = get_field_in("pseudo_density").get_view<const Pack**>();
+  auto        diag_view          = m_diagnostic_output.get_view<Pack**>();
+
+  const std::string diag_name           = m_diag_name;
+  const bool        from_sea_level      = m_from_sea_level;
+  const bool        is_interface_layout = m_is_interface_layout;
+  const int         num_levs            = m_num_levs;
+  auto              z_int               = m_z_int;
+  auto              z_mid               = m_z_mid;
+
+  Kokkos::parallel_for("VerticalLayerDiagnostic",
+                       default_policy,
+                       KOKKOS_LAMBDA(const MemberType& team) {
+    const int icol = team.league_rank();
+
+    // Calculate dz. Use the memory in z_mid for dz, since we don't
+    // set z_mid until after dz is no longer needed.
+    const auto& dz_s = ekat::subview(z_mid, icol);
+    Kokkos::parallel_for(Kokkos::TeamVectorRange(team, npacks), [&] (const Int& jpack) {
+      dz_s(jpack) = PF::calculate_dz(pseudo_density_mid(icol,jpack), p_mid(icol,jpack), T_mid(icol,jpack), qv_mid(icol,jpack));
+    });
+    team.team_barrier();
+
+    if (diag_name != "dz") {
+      // Calculate z_int if this diagnostic is not dz
+      const auto& z_int_s = ekat::subview(z_int, icol);
+      const Real surf_geopotential = from_sea_level ? 0.0 : phis(icol);
+      PF::calculate_z_int(team,num_levs,dz_s,surf_geopotential,z_int_s);
+
+      if (not is_interface_layout) {
+        // Calculate z_mid if this diagnostic is not dz or an interface value
+        const auto& z_mid_s = ekat::subview(z_mid, icol);
+        PF::calculate_z_mid(team,num_levs,z_int_s,z_mid_s);
+      }
+    }
+  });
+
+  // Copy data to the diagnostic output. For dz, z_mid, and geopotential_mid,
+  // z_mid contains the correct values. For z_int and geopotential_int, z_int.
+  if (m_is_interface_layout) Kokkos::deep_copy(diag_view, z_int);
+  else                       Kokkos::deep_copy(diag_view, z_mid);
+
+  const auto ts = get_field_in("qv").get_header().get_tracking().get_time_stamp();
```

Comment:
The base class already takes care of setting the timestamp in the diag field. so this is not needed.